dask/dataframe/io/json.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import absolute_import
+
+import io
+import pandas as pd
+from dask.bytes import open_files, read_bytes
+import dask
+
+
+def to_json(df, url_path, orient='records', lines=True, storage_options=None,
+            compute=True, **kwargs):
+    """Write dataframe into JSON text files
+
+    This utilises ``pandas.DataFrame.to_json()``, and most parameters are
+    passed through - see its docstring.
+
+    Differences: orient is 'records' by default, with lines=True; this
+    produces the kind of JSON output that is most common in big-data
+    applications, and which can be chunked when reading (see ``read_json()``).
+
+    Parameters
+    ----------
+    df: dask.DataFrame
+        Data to save
+    url_path: str, list of str
+        Location to write to. If a string, and there are more than one
+        partitions in df, should include a glob character to expand into a
+        set of file names, or provide a ``name_function=`` parameter.
+        Supports protocol specifications such as ``"s3://"``.
+    encoding, errors:
+        The text encoding to implement, e.g., "utf-8" and how to respond
+        to errors in the conversion (see ``str.encode()``).
+    orient, lines, kwargs
+        passed to pandas
+    storage_options: dict
+        Passed to backend file-system implementation
+    compute: bool
+        If true, immediately executes. If False, returns a set of delayed
+        objects, which can be computed at a later time.
+    """
+    kwargs['orient'] = orient
+    kwargs['lines'] = lines and orient == 'records'
```

Comment:
Are there options that the user might select that aren't valid that we don't check for?  I'm thinking about the option options to `lines` and `orient` here.  What happens if they choose something that is hard to do in parallel?

Reply:
Anything except orient='records' and lines=True does not parallelise - you get one partition per file. An attempt to set blocksize for those cases is an error. The reason for the `and` here, if that lines=True doesn't make sense for any other orient anyway, and would be an error in the pandas method.

Reply:
Do we give users a sensible error immediately, or does this error only happen at compute time?

If these are the only two options then we might consider just removing them as keywords.

Reply:
OK, if we leave them here then can we provide informative errors at graph construction time?  We've gotten complaints from downstream users whenever we don't fail early.  Apparently it's often quite hard to track down exactly which line in a complex analysis caused the error.  Perhaps something like the following:

```python
if orient not in {'records'}:
    raise NotImplementedError("The only valid value for the orient= keyword is 'records'")
if not lines:
    raise NotImplementedError("The lines= keyword must be set to True")
```

Reply:
I think you misunderstand me. Other orient values are fine and work (see the tests). The only difference is the choice of the defaults, since the ones here allow for parallel chunkwise reading later, and reflect the standards in spark/m-r.
The only thing that is not allowed by pandas (or, rather, does not make sense) is lines=True with any other orient.

Reply:
```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'x': [1, 2, 3]})

In [3]: for orient in {'split','records','index','columns','values'}:
   ...:     print(orient, df.to_json(orient=orient))
   ...:     
split {"columns":["x"],"index":[0,1,2],"data":[[1],[2],[3]]}
records [{"x":1},{"x":2},{"x":3}]
values [[1],[2],[3]]
columns {"x":{"0":1,"1":2,"2":3}}
index {"0":{"x":1},"1":{"x":2},"2":{"x":3}}
```

My sense here is that only records make sense if we're going to write data so that it can be split into many blocks within many files. 

However, I see now from reading the tests that we write many files that are independently readable with `pd.read_json` and ignore the blocksize keyword.  I guess this makes sense.  I'll withdraw most of this concern.

I do persist in the following request though:

> OK, if we leave them here then can we provide informative errors at graph construction time? We've gotten complaints from downstream users whenever we don't fail early. Apparently it's often quite hard to track down exactly which line in a complex analysis caused the error.

Reply:
If I do something like `df.to_json(orient='columns', lines=True, compute=False)` do I get an error immediately?

Reply:
This is not an error, and just works - the lines argument is ignored for all orients except records. It *could* raise an error instead, but I wouldn't like to do that, since `lines=True` by default.

Reply:
Alternatively we could also raise an informative error encouraging the user to use `orient='records'` or set `lines=False`.  I mention this only because many other people been upset with me in the past for implementing behaviors that might be surprising, even in what I considered to be corner-cases.

Reply:
The only potentially surprising behaviour would be that `to_json(orient='records')` (or no parameters at all) produces line-delimited output. The documentation on it I would say is pretty clear.
Another case like `to_json(orient='split')` works as expected, and `to_json(orient='split', lines=True)` also works where for pandas it would be an error - so no unexpected behaviour here.

Reply:
I don't personally have much experience with this method though, so my personal knowledge of what is best here is sparse.

Reply:
> to_json(orient='split', lines=True) also works where for pandas it would be an error - so no unexpected behaviour here.

I think the surprise here is that it doesn't produce a line-delimited output as was explicitly requested by the user.  I think that in this situation the correct thing to do is to raise an exception.

Reply:
How about `lines=None`, and the default for 'records' is `True`, but for everything else `False`, and in this case an error if `True` is selected for anything else, before passing on to pandas.

Reply:
Yes, that would satisfy my constraints.